File: handlers/investment.go
package handlers

import (
	"net/http"
	"strconv"

	"wealthwatch/middleware"
	"wealthwatch/models"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

type InvestmentHandler struct {
	db *gorm.DB
}

func NewInvestmentHandler(db *gorm.DB) *InvestmentHandler {
	return &InvestmentHandler{db: db}
}

type CreateHoldingRequest struct {
	AccountID      uint                 `json:"account_id" binding:"required"`
	Symbol         string               `json:"symbol" binding:"required"`
	Name           string               `json:"name" binding:"required"`
	InvestmentType models.InvestmentType `json:"investment_type" binding:"required"`
	Quantity       float64              `json:"quantity" binding:"required"`
	CostBasis      float64              `json:"cost_basis" binding:"required"`
	CurrentPrice   float64              `json:"current_price" binding:"required"`
}

func (h *InvestmentHandler) CreateHolding(c *gin.Context) {
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	var req CreateHoldingRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var account models.Account
	if err := h.db.Where("id = ? AND family_id = ?", req.AccountID, familyID).First(&account).Error; err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid account"})
		return
	}

	currentValue := req.Quantity * req.CurrentPrice
	gainLoss := currentValue - req.CostBasis
	var gainLossPercent float64
	if req.CostBasis > 0 {
		gainLossPercent = (gainLoss / req.CostBasis) * 100
	}

	holding := models.InvestmentHolding{
		AccountID:       req.AccountID,
		FamilyID:        familyID,
		Symbol:          req.Symbol,
		Name:            req.Name,
		InvestmentType:  req.InvestmentType,
		Quantity:        req.Quantity,
		CostBasis:       req.CostBasis,
		CurrentPrice:    req.CurrentPrice,
		CurrentValue:    currentValue,
		GainLoss:        gainLoss,
		GainLossPercent: gainLossPercent,
		LastUpdatedAt:   parseToday(),
	}

	if err := h.db.Create(&holding).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create holding"})
		return
	}

	c.JSON(http.StatusCreated, holding)
}

func (h *InvestmentHandler) ListHoldings(c *gin.Context) {
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	var holdings []models.InvestmentHolding
	q := h.db.Preload("Account").Where("family_id = ?", familyID).Order("current_value DESC")

	if accountID := c.Query("account_id"); accountID != "" {
		q = q.Where("account_id = ?", accountID)
	}
	if investmentType := c.Query("type"); investmentType != "" {
		q = q.Where("investment_type = ?", investmentType)
	}

	if err := q.Find(&holdings).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch holdings"})
		return
	}

	c.JSON(http.StatusOK, holdings)
}

func (h *InvestmentHandler) GetPortfolioSummary(c *gin.Context) {
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	var holdings []models.InvestmentHolding
	if err := h.db.Where("family_id = ?", familyID).Find(&holdings).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch holdings"})
		return
	}

	var totalValue, totalCostBasis, totalGainLoss float64
	allocationByType := map[string]float64{}

	for _, h := range holdings {
		totalValue += h.CurrentValue
		totalCostBasis += h.CostBasis
		totalGainLoss += h.GainLoss
		allocationByType[string(h.InvestmentType)] += h.CurrentValue
	}

	var totalGainLossPercent float64
	if totalCostBasis > 0 {
		totalGainLossPercent = (totalGainLoss / totalCostBasis) * 100
	}

	allocationPercent := map[string]float64{}
	if totalValue > 0 {
		for k, v := range allocationByType {
			allocationPercent[k] = (v / totalValue) * 100
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"total_value":          totalValue,
		"total_cost_basis":     totalCostBasis,
		"total_gain_loss":      totalGainLoss,
		"total_gain_loss_pct":  totalGainLossPercent,
		"allocation_by_type":   allocationByType,
		"allocation_percent":   allocationPercent,
		"holding_count":        len(holdings),
	})
}

type UpdateHoldingRequest struct {
	CurrentPrice *float64 `json:"current_price"`
	Quantity     *float64 `json:"quantity"`
}

func (h *InvestmentHandler) UpdateHolding(c *gin.Context) {
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid holding ID"})
		return
	}

	var holding models.InvestmentHolding
	if err := h.db.Where("id = ? AND family_id = ?", uint(id), familyID).First(&holding).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Holding not found"})
		return
	}

	var req UpdateHoldingRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if req.CurrentPrice != nil {
		holding.CurrentPrice = *req.CurrentPrice
	}
	if req.Quantity != nil {
		holding.Quantity = *req.Quantity
	}

	holding.CurrentValue = holding.Quantity * holding.CurrentPrice
	holding.GainLoss = holding.CurrentValue - holding.CostBasis
	if holding.CostBasis > 0 {
		holding.GainLossPercent = (holding.GainLoss / holding.CostBasis) * 100
	}
	holding.LastUpdatedAt = parseToday()

	if err := h.db.Save(&holding).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update holding"})
		return
	}

	c.JSON(http.StatusOK, holding)
}

func (h *InvestmentHandler) DeleteHolding(c *gin.Context) {
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid holding ID"})
		return
	}

	var holding models.InvestmentHolding
	if err := h.db.Where("id = ? AND family_id = ?", uint(id), familyID).First(&holding).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Holding not found"})
		return
	}

	if err := h.db.Delete(&holding).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete holding"})
		return
	}

	c.Status(http.StatusNoContent)
}

File: app/routers/investment.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import TokenData, get_current_user
from app.database import get_db
from app.models import InvestmentHolding, InvestmentType

router = APIRouter(prefix="/api/v1/investments", tags=["investments"])


class CreateHoldingRequest(BaseModel):
    account_id: int
    symbol: str
    name: str
    investment_type: str
    quantity: float = 0
    cost_basis: float = 0
    current_price: float = 0


class UpdateHoldingRequest(BaseModel):
    quantity: Optional[float] = None
    current_price: Optional[float] = None
    cost_basis: Optional[float] = None


def _holding_dict(h: InvestmentHolding) -> dict:
    return {
        "id": h.id, "account_id": h.account_id, "family_id": h.family_id,
        "symbol": h.symbol, "name": h.name, "investment_type": h.investment_type.value,
        "quantity": h.quantity, "cost_basis": h.cost_basis,
        "current_price": h.current_price, "current_value": h.current_value,
        "gain_loss": h.gain_loss, "gain_loss_percent": h.gain_loss_percent,
        "last_updated_at": str(h.last_updated_at) if h.last_updated_at else None,
        "created_at": str(h.created_at), "updated_at": str(h.updated_at),
    }


@router.get("")
async def list_holdings(
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    rows = (await db.execute(
        select(InvestmentHolding).where(InvestmentHolding.family_id == current_user.family_id)
        .order_by(InvestmentHolding.symbol)
    )).scalars().all()
    return [_holding_dict(h) for h in rows]


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_holding(
    req: CreateHoldingRequest,
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    current_value = req.quantity * req.current_price
    total_cost = req.quantity * req.cost_basis
    gain_loss = current_value - total_cost
    gain_loss_pct = (gain_loss / total_cost * 100) if total_cost else 0

    h = InvestmentHolding(
        account_id=req.account_id, family_id=current_user.family_id,
        symbol=req.symbol, name=req.name, investment_type=InvestmentType(req.investment_type),
        quantity=req.quantity, cost_basis=req.cost_basis, current_price=req.current_price,
        current_value=current_value, gain_loss=gain_loss, gain_loss_percent=gain_loss_pct,
    )
    db.add(h)
    await db.commit()
    await db.refresh(h)
    return _holding_dict(h)


@router.get("/portfolio")
async def get_portfolio_summary(
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    rows = (await db.execute(
        select(InvestmentHolding).where(InvestmentHolding.family_id == current_user.family_id)
    )).scalars().all()
    total_value = sum(h.current_value for h in rows)
    total_cost = sum(h.quantity * h.cost_basis for h in rows)
    by_type: dict[str, float] = {}
    for h in rows:
        by_type[h.investment_type.value] = by_type.get(h.investment_type.value, 0) + h.current_value
    return {
        "total_value": total_value, "total_cost": total_cost,
        "total_gain_loss": total_value - total_cost,
        "total_gain_loss_percent": ((total_value - total_cost) / total_cost * 100) if total_cost else 0,
        "by_type": by_type, "holding_count": len(rows),
    }


@router.put("/{id}")
async def update_holding(
    id: int, req: UpdateHoldingRequest,
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    h = (await db.execute(
        select(InvestmentHolding).where(InvestmentHolding.id == id, InvestmentHolding.family_id == current_user.family_id)
    )).scalar_one_or_none()
    if not h:
        raise HTTPException(404, "Holding not found")
    if req.quantity is not None:
        h.quantity = req.quantity
    if req.current_price is not None:
        h.current_price = req.current_price
    if req.cost_basis is not None:
        h.cost_basis = req.cost_basis
    h.current_value = h.quantity * h.current_price
    total_cost = h.quantity * h.cost_basis
    h.gain_loss = h.current_value - total_cost
    h.gain_loss_percent = (h.gain_loss / total_cost * 100) if total_cost else 0
    await db.commit()
    await db.refresh(h)
    return _holding_dict(h)


@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_holding(
    id: int,
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    h = (await db.execute(
        select(InvestmentHolding).where(InvestmentHolding.id == id, InvestmentHolding.family_id == current_user.family_id)
    )).scalar_one_or_none()
    if not h:
        raise HTTPException(404, "Holding not found")
    await db.delete(h)
    await db.commit()
